File: 3123. Find Edges in Shortest Paths.py
class Solution:
    def findAnswer(self, n: int, edges: List[List[int]]) -> List[bool]:
        graph = {}
        for edge in edges:
            if edge[0] not in graph:
                graph[edge[0]] = []
            if edge[1] not in graph:
                graph[edge[1]] = []
            graph[edge[0]].append([edge[1], edge[2]])
            graph[edge[1]].append([edge[0], edge[2]])
        def find_distance(source, target):
            nonlocal graph, n
            distance = [int (1e9)] * n
            distance[source] = 0
            visited = set()
            pq = [(0, source)]
            while pq:
                dist, node = heapq.heappop(pq)
                if node in visited:
                    continue
                visited.add(node)
                if node not in graph:
                    continue
                for neighbor, weight in graph[node]:
                    if distance[neighbor] > distance[node] + weight:
                        distance[neighbor] = distance[node] + weight
                        heapq.heappush(pq, (distance[neighbor], neighbor))
            return distance
        distance_source = find_distance(0, n-1)
        distance_target = find_distance(n-1, 0)
        answer = [False] * len(edges)
        for i in range(len(edges)):
            u, v, w = edges[i]
            if distance_source[u] + w + distance_target[v] == distance_source[n-1]:
                answer[i] = True
            if distance_source[v] + w + distance_target[u] == distance_source[n-1]:
                answer[i] = True
        return answer

File: 3123. Find Edges in Shortest Paths.go
func findAnswer(n int, edges [][]int) []bool {
	graph := make(map[int][][2]int)
	for _, edge := range edges {
		u, v, w := edge[0], edge[1], edge[2]
		graph[u] = append(graph[u], [2]int{v, w})
		graph[v] = append(graph[v], [2]int{u, w})
	}
	distanceSource := findDistances(graph, 0, n)
	distanceTarget := findDistances(graph, n-1, n)
	answer := make([]bool, len(edges))
	for i, edge := range edges {
		u, v, w := edge[0], edge[1], edge[2]
		if distanceSource[u]+w+distanceTarget[v] == distanceSource[n-1] {
			answer[i] = true
		}
		if distanceSource[v]+w+distanceTarget[u] == distanceSource[n-1] {
			answer[i] = true
		}
	}
	return answer
}

func findDistances(graph map[int][][2]int, source, n int) []int {
	distance := make([]int, n)
	for i := range distance {
		distance[i] = math.MaxInt32
	}
	distance[source] = 0
	pq := make(PriorityQueue, 0)
	heap.Push(&pq, &Node{source, 0})
	for pq.Len() > 0 {
		node := heap.Pop(&pq).(*Node)
		u := node.val
		for _, edge := range graph[u] {
			v, w := edge[0], edge[1]
			if distance[u]+w < distance[v] {
				distance[v] = distance[u] + w
				heap.Push(&pq, &Node{v, distance[v]})
			}
		}
	}
	return distance
}

type Node struct {
	val      int
	distance int
}

type PriorityQueue []*Node

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool { return pq[i].distance < pq[j].distance }

func (pq PriorityQueue) Swap(i, j int) { pq[i], pq[j] = pq[j], pq[i] }

func (pq *PriorityQueue) Push(x interface{}) {
	node := x.(*Node)
	*pq = append(*pq, node)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}